packages/arb-avm-cpp/cmachine/machine.go

```diff
@@ -194,12 +203,34 @@ func (m *Machine) ExecuteAssertionAdvanced(
 
 	C.machineExecutionConfigSetStopOnSideload(conf, boolToCInt(stopOnSideload))
 
-	assertionResult := C.executeAssertion(m.c, conf)
+	resultChan := make(chan C.RawAssertionResult, 1)
+	go func() {
+		defer close(resultChan)
+		resultChan <- C.executeAssertion(m.c, conf)
+	}()
+
+	aborted := false
+	var assertionResult C.RawAssertionResult
+	select {
+	case <-ctx.Done():
+		m.Abort()
+		aborted = true
+		// Still need to clean up result, so wait until returned
+		assertionResult = <-resultChan
+	case assertionResult = <-resultChan:
+	}
+
+	// Make sure result is cleaned up properly
+	executionAssertion, values, steps, err := makeExecutionAssertion(assertionResult.assertion)
+
+	if aborted {
+		return nil, nil, 0, errors.New("aborted")
```

Comment:
Returning `ctx.Err()` here should give a normal context canceled error, which is probably a bit cleaner.

Reply:
:+1: